compiler/rustc_trait_selection/src/traits/outlives_bounds.rs

```diff
@@ -52,6 +53,10 @@ impl<'a, 'tcx: 'a> InferCtxtExt<'a, 'tcx> for InferCtxt<'tcx> {
         body_id: LocalDefId,
         ty: Ty<'tcx>,
     ) -> Vec<OutlivesBound<'tcx>> {
+        let ty = self.resolve_vars_if_possible(ty);
+        let ty = OpportunisticRegionResolver::new(self).fold_ty(ty);
```

Comment:
Why do we need to resolve infer vars here? cc https://github.com/rust-lang/rust/issues/112832

Reply:
to avoid any assumptions using inference vars. E.g. if you have `'a = '2` and then get an implied bound `'2: 'b`, then getting `'a: 'b` from that requires us to know that `'a = '2` is something we can assume and not something we have to prove.

All region vars which are part of the tpye before calling `implied_outlives_bounds` are from candidates used while normalizing the `ty`, so we can assume that equality holds there.

We need to make sure that we don't use a region bound we have to prove to strengthen an assumption. Had an example for that at some point, would have to look for it though.